### PR TITLE
Types refactor

### PR DESCRIPTION
**Pull request by rafael-atticlabs**
Review this PR in patch order. 

1. Decoder no longer needed to remove struct cycles. That happened as we decoded
2. There were a bunch of tests that were asserting something about structs after removing cycles. I suspect these tests did something useful at some point, but I'm pretty they weren't any more. Apologies if I got this wrong. If so, assign bug to me to restore anything important.
3. Remove dead code
4. Refactor and add test.
5. Just moving code around (mostly type_cache -> make_type)
6. Remove (no longer neccesary) typeDepth counter in ValueDecoder 

TBR @arv

### CODE REVIEW

File: go/types/simplify_test.go [hunk 1]
```diff
@@ -64,20 +64,6 @@ func TestSimplifyType(t *testing.T) {
 			act := simplifyType(in, intersectStructs)
 			assert.True(exp.Equals(act), "Expected: %s\nActual: %s", exp.Describe(), act.Describe())
 		}
-		testSame := func(t *Type) {
-			test(t, resolveCycleTypes(t))
-		}
-
-		testSame(BlobType)
```
Comment (arv) [inline go/types/simplify_test.go]: These were useful. It ensures that we don't do anything stupid.

File: go/types/type.go [hunk 2]
```diff
@@ -111,41 +111,33 @@ func TypeOf(v Value) *Type {
 
 // HasStructCycles determines if the type contains any struct cycles.
 func HasStructCycles(t *Type) bool {
-	return hasStructCycles(t, map[string]int{})
+	return hasStructCycles(t, nil)
 }
 
-func hasStructCycles(t *Type, seenStructs map[string]int) bool {
+func hasStructCycles(t *Type, visited []*Type) bool {
```
Comment (arv) [inline go/types/type.go]: Not sure this is better? It comes down to higher constant factor O(1) vs lower constant factor O(n).

But lets keep the slice solution. It is simpler to understand.

File: go/types/simplify_test.go (hunk 1 above)
Comment (rafael-atticlabs) [inline go/types/simplify_test.go]: I'm not sure I understand. They were tests that we didn't do random bad stuff to types?

**Comment by arv** [comment]
No they were testing that the output of simplify was what we expect.

**Comment by rafael-atticlabs** [comment]
Ah. I get it now. Sorry. I'll put them back (without resolveCycles) tomorrow.

**Pull request merged**